proposals/0000-unstable.rst

```diff
@@ -0,0 +1,306 @@
+available; this knowledge can help inform the design of the feature. An excellent
+example is ``-XOverloadedRecordDot``. This extension was essentially implemented
+within Digital Asset's internal version of Haskell for some time before being proposed in GHC.
+Through years of experimentation and user feedback, the design was refined. GHC then
+benefited from this history in getting a new, yet battle-tested extension. If Digital
+Asset had not invested in this way, though, we would have no way of gaining experience
+without trapping ourselves in a potentially backward-incompatible corner. With ``-XUnstable``,
+we might be able to roll out more features, learn from their use, and then refine.
+
+``-XStable``
+~~~~~~~~~~~~
+
+Unstable features are, well, unstable. Consequently, some developers and enterprises
+may wish to forbid these features in their code. It works best to have an automated
+way of enforcing this prohibition. The ``-XStable`` extension does exactly this: it prevents
+the use of ``-XUnstable`` (only). Now, a developer might put ``-XStable`` in its ``default-extensions``
+field of a ``.cabal`` file to be sure that a package uses no unstable features of GHC.
+Perhaps even more powerfully, a potential client perusing libraries on Hackage can view
+``.cabal`` files and choose to use only those that have ``-XStable`` in their ``.cabal`` file,
+suggesting that these libraries are more likely to continue compiling in future GHC versions.
+
+Proposed Change Specification
+-----------------------------
+
+There are two components of this proposal: the ``-XUnstable`` extension and the
+``-XStable`` extension. The components are separable (``-XUnstable`` makes sense
+even without ``-XStable``), but ``-XStable`` depends on ``-XUnstable``.
+
+``-XUnstable``
+~~~~~~~~~~~~~~
+
+Technical specification
+#######################
+
+1. Introduce a new extension ``-XUnstable``. This extension is off by default and
+   not a part of any extension set.
+
+#. When a module from the ``ghc-prim`` package is imported, if ``-XUnstable`` is not
+   specified, report an error informing the user that modules from ``ghc-prim`` are
+   unstable. After warning the user of the consequences, the error would inform users
+   that specifying ``-XUnstable`` makes the error go away.
+
+#. Remove ``FUN`` from the export list of ``GHC.Exts`` (so that ``FUN`` becomes unstable).
+
+#. (Technically beyond the scope of this proposal process) The module-level documentation
+   for each of the exported modules of ``ghc-prim`` will include a note that users
+   should not import that module directly; instead, users will be directed where to
+   find the imports they need.
+
+#. (Technically beyond the scope of this proposal process) Tooling that helps
+   users automatically insert extensions (i.e. HLS) would not automatically enable
+   ``-XUnstable`` from this message; users would have to add it manually.
```

Comment:
What is itabout `-XUnstable` that makes it deserve that treatment? Are there other extensions that should not be “easily enabled”? (Maybe `Trustworthy`, `IncoherentInstances`)?

And is it even good style? If the user already added `ghc-prim` to their cabal file, and added the imports, they are very likely determined to use `ghc-prim` and hopefully know why. Why make them do manual busy work that the tool could do for them?

(Yes, it’s not technically part of the proposal and the HLS authors might just choose to ignore it, or implement it by allowing automatic after a warning.)